aws/data_source_aws_rds_orderable_db_instance.go

```diff
@@ -0,0 +1,368 @@
+
+		if output == nil {
+			break
+		}
+
+		for _, instanceOption := range output.OrderableDBInstanceOptions {
+			if instanceOption == nil {
+				continue
+			}
+
+			if v, ok := d.GetOk("engine_version_prefix"); ok {
+				engineVersion := aws.StringValue(instanceOption.EngineVersion)
+				if !strings.HasPrefix(engineVersion, v.(string)) {
+					continue
+				}
+			}
+
+			if v, ok := d.GetOk("storage_type"); ok {
+				if aws.StringValue(instanceOption.StorageType) != v.(string) {
+					continue
+				}
+			}
+
+			instanceClass := aws.StringValue(instanceOption.DBInstanceClass)
+			instanceClassResults = append(instanceClassResults, instanceClass)
+			instanceInfo[instanceClass] = instanceOption
+
+			instanceEngineVersions = append(instanceEngineVersions, aws.StringValue(instanceOption.EngineVersion))
+		}
+
+		if aws.StringValue(output.Marker) == "" {
+			break
+		}
+
+		input.Marker = output.Marker
+	}
+
+	if len(instanceClassResults) == 0 {
+		return fmt.Errorf("no RDS Orderable DB Instance options found matching criteria; try different search")
+	}
+
+	// if prefix is used, limit results to highest version engines in the results
+	highestVersion := ""
+	if v, ok := d.GetOk("engine_version_prefix"); ok {
+		for _, version := range instanceEngineVersions {
+			highestVersion = dataSourceAwsRdsOrderableDbInstanceReadVersionCompare(
+				v.(string),
+				version,
+				highestVersion,
+			)
+		}
+
+		if highestVersion != "" {
+			var newInstanceClassResults []string
+			for i, instanceClassResult := range instanceClassResults {
+				if instanceEngineVersions[i] != highestVersion {
+					continue
+				}
+				newInstanceClassResults = append(newInstanceClassResults, instanceClassResult)
+			}
+			instanceClassResults = newInstanceClassResults
+		}
+	}
+
+	// preferred classes
+	var foundInstanceClass string
+	if l := d.Get("preferred_db_instance_classes").([]interface{}); len(l) > 0 {
+		for _, elem := range l {
+			preferredInstanceClass, ok := elem.(string)
+
+			if !ok {
+				continue
+			}
+
+			for _, instanceClassResult := range instanceClassResults {
+				if instanceClassResult == preferredInstanceClass {
+					foundInstanceClass = preferredInstanceClass
+					break
+				}
+			}
+
+			if foundInstanceClass != "" {
+				break
+			}
+		}
+	}
+
+	if foundInstanceClass == "" && len(instanceClassResults) > 1 {
+		return fmt.Errorf("multiple RDS DB Instance Classes (%v) match the criteria; try a different search", instanceClassResults)
+	}
+
+	if foundInstanceClass == "" && len(instanceClassResults) == 1 {
+		foundInstanceClass = instanceClassResults[0]
+	}
+
+	if foundInstanceClass == "" {
+		return fmt.Errorf("no RDS DB Instance Classes match the criteria; try a different search")
+	}
+
+	d.SetId(resource.UniqueId())
+
+	d.Set("db_instance_class", foundInstanceClass)
+
+	instanceOption := *instanceInfo[foundInstanceClass].(*rds.OrderableDBInstanceOption)
+
+	d.Set("availability_zone_group", aws.StringValue(instanceOption.AvailabilityZoneGroup))
```

Comment:
Nit: Here and below, `d.Set()` automatically handles pointers, so we don't need to add the conversion functions. This will ensure that references will be missing for things the API doesn't return.

```suggestion
	d.Set("availability_zone_group", instanceOption.AvailabilityZoneGroup)
```